### DISCUSSION TITLE
Apply style to Border#PART_ContentPresenterBorder of a TextBox

### DISCUSSION THREAD

**Discussion by DevAM-Tools** [AvaloniaUI/Avalonia#15627] [Q&A]
I want to minimize the size of a `TextBox` when used with a small font size. I noticed that a `Border#PART_ContentPresenterBorder` within the visual tree of the TextBox has a `MinHeight` set. I want to set this to 0 for every `TextBox`.

I came along with the following style. But this has the disadvantage that every `TextBox` needs to get the `Class` `Small` assigned to it to apply the style. Is there a way to apply this styling to every `TextBox` without setting the `Class`?

```Xaml
<Style Selector="TextBox.Small /template/ Border#PART_ContentPresenterBorder">
	<Setter 
		Property="MinHeight" 
		Value="0" />
</Style>
```

**Comment by thevortexcloud** [comment]
You should be able to apply the style in your application styles. However it may not work depending on which style is more specific, I am not sure what Avalonia's precedence rules are in this case. If it does not work, a bruteforce work around is to just override the template with a custom one.

**Reply by DevAM-Tools** [reply]
If I use `TextBox.Small` as selector the DevTools show `StyleTrigger` as priority.
```Xaml
<Style Selector="TextBox.Small /template/ Border#PART_ContentPresenterBorder">
	<Setter 
		Property="MinHeight" 
		Value="0" />
</Style>
```

If I use `TextBox` as selector (without class `Small`) the DevTools show `Template` as Priority.
```Xaml
<Style Selector="TextBox /template/ Border#PART_ContentPresenterBorder">
	<Setter 
		Property="MinHeight" 
		Value="0" />
</Style>
```

So you are probably right. It seems to be releated to the precedence of styling rules somehow.

**Comment by robloo** [comment]
Which theme are you using? Fluent, Simple or a 3rd party?

You are referring to a `PART_ContentPresenterBorder` which is NOT in the templates for `TextBox` in any of our themes so either that's a mistake or you are using a 3rd party theme we need to know about. Also.. that template part is misnamed I think -- we really should have cleaned those up more.

One of the biggest issues in Avalonia styling is precedence with style selectors. These cannot override values set in control templates which have higher local priority (effectively making them useless in a lot of cases). Work arounds are to go deeper into the template and select on content presenter, etc. directly. See #7679 (reverted) and #13601.

However, in this case I don't think that's the issue. I looked at both Fluent and Simple theme (you have to tell us the theme you are using) and see that MinHeight on the border in set by a TemplateBinding so it is overridable just fine.

Long story short, for in-box Avalonia themes you should simply be using:

```xaml
<Style Selector="TextBox">
	<Setter 
		Property="MinHeight" 
		Value="0" />
</Style>
```

**Reply by DevAM-Tools** [reply]
Thank you for your explanation.

I am using `FluentTheme` with `Semi.Avalonia`
```Xaml
<Application.Styles>
    <FluentTheme />
    <StyleInclude Source="avares://Semi.Avalonia/Themes/Index.axaml" />
</Application.Styles>
```

The TextBox is basically just an example. I have the same issue also with other Controls. Based on your feedback my question could also be written as:

 _How do I overwrite a style set by a template without assigning a Class to every instance?_

**Reply by timunie** [reply]
I don't really suggest to mix these two styles unless Semi.Avalonia requires it.
